Interop: NuGet packages and other cleanup

RandomEngy opened the pull request:
Also added some constants for keep width/height that can be used in the picture settings for JSON.

RandomEngy committed:
Changed library references to use NuGet and did a bit more cleanup.

bradleysepos commented:
Cool.

Would be cleaner to rebase the branch on current master than merge master onto the branch.

RandomEngy commented:
Bah. Used to merging from TFS where it doesn't create another change when merging back. Do you need me to redo this? Not sure how to back out the merge on my current repo.

bradleysepos commented:
Assuming you have no other changes uncommitted, update your master branch, and then make a backup of your interop branch:

```
git checkout master
git pull --ff-only <upstream-remote> master
git checkout interop
git branch interop-backup
```

The master branch is now up to date with the official repo, and you're on the interop branch (`git branch` copied the branch but did not switch to the backup).

To rebase this _specific_ timeline, the following will work:

```
git reset --hard ab9a449
```

HEAD is now at ab9a449 "Updates to support getting valid filter parameters dynamically"

```
git rebase master
```

Edit .gitignore to resolve the marked conflicts. Basically, make the file look like it should look after all is said and done.

You will see an opening <<< line followed by the original file's contents, a === separator followed by your changes, and then a closing >>> line. It seems in this case you want to remove the conflicting line entirely, and of course the <<< === >>> lines. The .gitignore should now look like a normal .gitignore.

Use `git add` to mark the conflicts as resolved, then instruct git to continue rebasing.

```
git add .gitignore
git rebase --continue
```

Since the .gitignore ended up having no changes in the end, it's not in your final commit at all. You may want to edit your commit message to reflect this. Then push the changes to your fork on GitHub (force is needed since you're changing history). Your pull request should automatically reflect the changes.

```
git commit --amend
git push --force <your-remote-fork> interop
```

If you get stuck, you can always revert to the starting point and try again with a hard reset to that ugly merge commit:

```
git reset --hard e9c1946
```

Git will keep the old/undesired timeline hidden away until you delete the backup branch (`git branch -D interop-backup`) _and_ run some sort of garbage collection on the repo (`git gc`). You may want to do the former once you're satisfied. The latter may recover some space at the expense of permanently losing any discarded history; it's not necessary to perform garbage collection often, if ever.

sr55 commented on win/CS/HandBrake.ApplicationServices/Interop/HbLib/HbFunctions.cs:
```diff
@@ -362,6 +362,14 @@ public static class HBFunctions
 
+        ///hb_filter_param_t* hb_filter_param_get_presets(int filter_id);
+        [DllImport("hb.dll", EntryPoint = "hb_filter_param_get_presets", CallingConvention = CallingConvention.Cdecl)]
+        public static extern IntPtr hb_filter_param_get_presets(int filter_id);
```
The JSON version of these calls are already implemented. See hb_filter_get_tunes_json below.
HandBrakeFilterHelpers contains two methods that return the data in object form. Although the object name could do with being changed.

RandomEngy commented:
Okay, I've reworked this to no longer include the filter values (which were already implemented). I've changed to convert all external library references to NuGet. This has a few advantages:
1. Easier to update external libraries
2. I've added the packages folder to .gitignore so it relies on NuGet package restore and does not need to include the DLLs in the repo, making it lighter weight.

I also did a rebase and squashed down the commits to one as requested.
